ADC: STM32: Add device tree config for clock source with prescaler

FRASTM opened the pull request:
This PR adds a clock prescaler to the ADC of the stm32 mcus

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/40955

Signed-off-by: Francois Ramu <francois.ramu@st.com>

Commit:
dts: bindings: adc stm32 has a prescaler entry for the ADC clock

This is a new parameter to divide the ADC clock by the ADC_PRESC.
The ADC clock is CLOCK_SYNC_PCLK or CLOCK_ASYNC depending on the serie.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

pdietl commented on drivers/adc/adc_stm32.c:
```diff
@@ -236,6 +236,45 @@ static const uint32_t table_samp_time[] = {
+};
+
+/* Bugfix for STM32G4 HAL */
+#if !defined(ADC_CHANNEL_TEMPSENSOR)
+#define ADC_CHANNEL_TEMPSENSOR ADC_CHANNEL_TEMPSENSOR_ADC1
+#endif
```
Why is this added back in?

FRASTM replied:
removing this code

pdietl commented on dts/bindings/adc/st,stm32-adc.yaml:
```diff
@@ -27,6 +27,14 @@ properties:
     pinctrl-names:
       required: true
 
+    st,prescaler:
+      type: int
+      required: false
+      default: 0
+      description: |
+        Clock prescaler at the input of the adc
+        Could be in range [0 .. 0xF] ; default is 0 (not divided)
+
```
Does it make sense to use an enum type here?

FRASTM replied:
adding `Valid range: 0 - 15`

FRASTM replied:
adding `Valid range: 0 - 15`

mbolivar-nordic commented on dts/bindings/adc/st,stm32-adc.yaml:
```diff
@@ -27,6 +27,14 @@ properties:
+      default: 0
+      description: |
+        Clock prescaler at the input of the adc
+        Could be in range [0 .. 0xF] ; default is 0 (not divided)
```
This default needs a justification as per https://docs.zephyrproject.org/latest/guides/dts/bindings.html#default-values-for-properties

FRASTM replied:
Ok, default is the reset value

pdietl commented on drivers/adc/adc_stm32.c:
```diff
@@ -236,6 +236,41 @@ static const uint32_t table_samp_time[] = {
 };
 #endif
 
+/* table to define the LL ADC clock prescaler from the DTS */
```
We should add a static assert that `config->clock_prescaler < ARR_SIZE(adc_clock_prescaler)`

FRASTM replied:
with ASSERT_NO_MSG

pdietl commented on drivers/adc/adc_stm32.c:
```diff
@@ -921,6 +957,9 @@ static int adc_stm32_init(const struct device *dev)
 		return -EIO;
 	}
 
+	__ASSERT_NO_MSG(config->clock_prescaler < ARRAY_SIZE(adc_clock_prescaler));
```
This is still a runtime assert. This could be found at compile time.

FRASTM replied:
changing to BUILD_ASSERT

erwango commented on dts/bindings/adc/st,stm32-adc.yaml:
```diff
@@ -27,6 +27,31 @@ properties:
+        - 15
+      description: |
+        Clock prescaler at the input of the adc
+        Could be in range [0 .. 0xF] ; default reset value is 0 (not divided)
```
Remove "Could be in range [0 .. 0xF] ; "

FRASTM replied:
removed

erwango reviewed (changes requested):
Please update commit author: `author email (Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>) needs to match one of the signed-off-by entries.`

`git amend --author="François ...."`

FRASTM commented:
This part of the stm32 adc driver in the `adc_stm32_init` function requires more changes because the ADC clock can now hav different prescaler value 
including Synchronous or asynchronous division factor. That should be handled a particular call to LL_ADC_SetCommonClock and/or LL_ADC_SetClock depending on the stm32 soc

FRASTM commented:
rebase on

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -236,6 +236,41 @@ static const uint32_t table_samp_time[] = {
+	LL_ADC_CLOCK_SYNC_PCLK_DIV4,
+#elif defined(CONFIG_SOC_SERIES_STM32L1X) || \
+	defined(CONFIG_SOC_SERIES_STM32U5X)
+	LL_ADC_CLOCK_ASYNC_DIV1,
+	LL_ADC_CLOCK_ASYNC_DIV2,
+	LL_ADC_CLOCK_ASYNC_DIV4,
+	LL_ADC_CLOCK_ASYNC_DIV6,
+	LL_ADC_CLOCK_ASYNC_DIV8,
+	LL_ADC_CLOCK_ASYNC_DIV10,
+	LL_ADC_CLOCK_ASYNC_DIV12,
+	LL_ADC_CLOCK_ASYNC_DIV16,
+	LL_ADC_CLOCK_ASYNC_DIV32,
+	LL_ADC_CLOCK_ASYNC_DIV64,
+	LL_ADC_CLOCK_ASYNC_DIV128,
+	LL_ADC_CLOCK_ASYNC_DIV256,
+#endif
```
These are also supported by most other series: CONFIG_SOC_SERIES_STM32L4X || CONFIG_SOC_SERIES_STM32L5X || CONFIG_SOC_SERIES_STM32WBX || CONFIG_SOC_SERIES_STM32G0X || CONFIG_SOC_SERIES_STM32G4X || CONFIG_SOC_SERIES_STM32H7X

FRASTM replied:
CONFIG_SOC_SERIES_STM32U5X and CONFIG_SOC_SERIES_STM32L0X only have LL_ADC_CLOCK_ASYNC_DIVx  for their common clock config. 
Soc series that have LL_ADC_CLOCK_SYNC_PCLK_DIVx and also LL_ADC_CLOCK_ASYNC_DIVx to set their LL_ADC_SetCommonClock, would require a prescaler for Synchronous clock and one for Asynchronous clock.

erwango reviewed (changes requested):
We need to revisit this change as it mixes CKMODE and PRESC and binding doesn't reflect sufficiently allow to describe socs capabilities/limitations

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -236,6 +236,41 @@ static const uint32_t table_samp_time[] = {
 };
 #endif
 
+/* table to define the LL ADC clock prescaler from the DTS */
+static const uint32_t adc_clock_prescaler[] = {
+#if defined(CONFIG_SOC_SERIES_STM32F0X) || \
+	defined(CONFIG_SOC_SERIES_STM32L0X) || \
+	defined(CONFIG_SOC_SERIES_STM32WLX)
+	LL_ADC_CLOCK_SYNC_PCLK_DIV1,
+	LL_ADC_CLOCK_SYNC_PCLK_DIV2,
+	LL_ADC_CLOCK_SYNC_PCLK_DIV4,
+#elif defined(STM32F3X_ADC_V1_1) || \
+	defined(CONFIG_SOC_SERIES_STM32L4X) || \
+	defined(CONFIG_SOC_SERIES_STM32L5X) || \
+	defined(CONFIG_SOC_SERIES_STM32WBX) || \
+	defined(CONFIG_SOC_SERIES_STM32G0X) || \
+	defined(CONFIG_SOC_SERIES_STM32G4X) || \
+	defined(CONFIG_SOC_SERIES_STM32H7X)
+	LL_ADC_CLOCK_SYNC_PCLK_DIV1,
+	LL_ADC_CLOCK_SYNC_PCLK_DIV2,
+	LL_ADC_CLOCK_SYNC_PCLK_DIV4,
+#elif defined(CONFIG_SOC_SERIES_STM32L1X) || \
+	defined(CONFIG_SOC_SERIES_STM32U5X)
+	LL_ADC_CLOCK_ASYNC_DIV1,
+	LL_ADC_CLOCK_ASYNC_DIV2,
+	LL_ADC_CLOCK_ASYNC_DIV4,
+	LL_ADC_CLOCK_ASYNC_DIV6,
+	LL_ADC_CLOCK_ASYNC_DIV8,
+	LL_ADC_CLOCK_ASYNC_DIV10,
+	LL_ADC_CLOCK_ASYNC_DIV12,
+	LL_ADC_CLOCK_ASYNC_DIV16,
+	LL_ADC_CLOCK_ASYNC_DIV32,
+	LL_ADC_CLOCK_ASYNC_DIV64,
+	LL_ADC_CLOCK_ASYNC_DIV128,
+	LL_ADC_CLOCK_ASYNC_DIV256,
+#endif
+};
+
```
```suggestion
/* table to define the LL ADC clock prescaler from the DTS */

#if defined(CONFIG_SOC_SERIES_STM32F0X) || \
	defined(CONFIG_SOC_SERIES_STM32L0X) || \
	defined(CONFIG_SOC_SERIES_STM32WLX)
        defined(STM32F3X_ADC_V1_1) || \
	defined(CONFIG_SOC_SERIES_STM32L4X) || \
	defined(CONFIG_SOC_SERIES_STM32L5X) || \
	defined(CONFIG_SOC_SERIES_STM32WBX) || \
	defined(CONFIG_SOC_SERIES_STM32G0X) || \
	defined(CONFIG_SOC_SERIES_STM32G4X) || \
	defined(CONFIG_SOC_SERIES_STM32H7X)
#define z_clk(v)	LL_ADC_CLOCK_SYNC_PCLK_##v
#elif defined(CONFIG_SOC_SERIES_STM32L1X) || \
	defined(CONFIG_SOC_SERIES_STM32U5X)
#define z_clk(v)	LL_ADC_CLOCK_ASYNC_##v
#endif

#define adc_clock_prescaler(v) z_clk(v)
```

Then you can do
```
LL_ADC_SetClock(adc, adc_clock_prescaler[config->clock_prescaler]);
```

str4t0m replied:
I think that specifying the actual clock-divider via dts instead of the register value would be more intuitive for the user.
And then you could use the macro recommended by erwango instead of using an array to translate the register value back and forth.

FRASTM replied:
Defining the st,prescaler property  directly with the divider value to be 1, 2, 4, 6, 8, 10, 12, 16, 32, 64,128,256  is possible in the dts. The related adc_clock_prescaler will depend on the soc serie 
either LL_ADC_CLOCK_SYNC_PCLK_##v or LL_ADC_CLOCK_ASYNC_##v 
Except for the **stm32F0** where default 1 is LL_ADC_CLOCK_ASYNC and 2 or 4 are LL_ADC_CLOCK_SYNC_PCLK_DIV_##v

FRASTM replied:
then configuring with  `LL_ADC_SetClock(adc, adc_clock_prescaler(config->clock_prescaler));`

erwango replied:
> However with` #define adc_clock_prescaler(v) z_clock_prescaler(v)` "v" must be like "DT_INST_PROP(0, st_prescaler)" and cannot be "config->clock_prescaler" (-->`error: 'LL_ADC_CLOCK_ASYNC_DIVconfig' undeclared`)

Right, since this rely on pre-processing macros 'v' should be available at build time, which is not the case for config->clock_prescaler.

Anyway, we should find another way to allow the user to provide the prescaler value instead of an array rank.

FRASTM replied:
Using COND_CODE_1, it is possible to build a LL_ADC_CLOCK_ASYNC_DIVn or LL_ADC_CLOCK_SYNC_PCLK_DIVn
value from the st,prescaler property (based on string concatenation)

FRASTM committed:
drivers: adc: stm32 adc has a prescaler to divide the ADC clock

This makes the selection of the ADC clock prescaler to divide
the ADC common clock thanks to the dts entry.
The prescaler value depends on the stm32 soc serie and fits the
LL constant to pass to the LL function.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented on drivers/adc/adc_stm32.c:
```diff
@@ -236,6 +236,41 @@ static const uint32_t table_samp_time[] = {
+#if defined(CONFIG_SOC_SERIES_STM32F0X) || \
+	defined(CONFIG_SOC_SERIES_STM32L0X) || \
+	defined(CONFIG_SOC_SERIES_STM32WLX)
+	LL_ADC_CLOCK_SYNC_PCLK_DIV1,
```
removed for stm32F0x

FRASTM commented on drivers/adc/adc_stm32.c:
```diff
@@ -236,6 +236,41 @@ static const uint32_t table_samp_time[] = {
+	defined(CONFIG_SOC_SERIES_STM32H7X)
+	LL_ADC_CLOCK_SYNC_PCLK_DIV1,
+	LL_ADC_CLOCK_SYNC_PCLK_DIV2,
+	LL_ADC_CLOCK_SYNC_PCLK_DIV4,
```
merged

FRASTM commented on drivers/adc/adc_stm32.c:
```diff
@@ -973,11 +1011,11 @@ static int adc_stm32_init(const struct device *dev)
-			      LL_ADC_CLOCK_SYNC_PCLK_DIV4);
+			      adc_clock_prescaler[config->clock_prescaler]);
 #elif defined(CONFIG_SOC_SERIES_STM32L1X) || \
 	defined(CONFIG_SOC_SERIES_STM32U5X)
```
removed

FRASTM commented:
- rebase on 5897de1e535816247e309348cbe29760361a28b8
- define the st,prescaler as a string and build the clock_prescaler using concatenation.

Example for stm32u5x, to set LL_ADC_CLOCK_ASYNC_DIV6, the dts must define the property as follows: 
```
&adc1 {
	st,prescaler = "6";
};
```

FRASTM commented:
The stm32F1x has no prescaler, its config->clock_prescaler is reset

FRASTM commented:
See https://github.com/zephyrproject-rtos/zephyr/pull/48938
